pygmt/src/tilemap.py

```diff
@@ -0,0 +1,145 @@
+"""
+tilemap - Plot XYZ tile maps.
+"""
+
+from pygmt.clib import Session
+from pygmt.helpers import (
+    GMTTempFile,
+    build_arg_string,
+    fmt_docstring,
+    kwargs_to_strings,
+    use_alias,
+)
+
+try:
+    import rioxarray
+except ImportError:
+    rioxarray = None
+
+
+@fmt_docstring
+@use_alias(
+    A="img_out",
+    B="frame",
+    C="cmap",
+    D="img_in",
+    E="dpi",
+    G="bit_color",
+    I="shading",
+    J="projection",
+    M="monochrome",
+    N="no_clip",
+    Q="nan_transparent",
+    # R="region",
+    V="verbose",
+    n="interpolation",
+    c="panel",
+    f="coltypes",
+    p="perspective",
+    t="transparency",
+    x="cores",
+)
+@kwargs_to_strings(c="sequence_comma", p="sequence")  # R="sequence",
+def tilemap(
+    self, region, zoom="auto", source=None, lonlat=True, wait=0, max_retries=2, **kwargs
+):
```

Comment:
Currently `region` needs to be a list to work with `load_tile_map`, but I'm debating on whether to allow strings like `0/1/2/3` or other styles supported by GMT `-R` as mentioned in https://github.com/GenericMappingTools/pygmt/issues/153#issuecomment-742896187 and https://docs.generic-mapping-tools.org/6.4/cookbook/options.html#data-domain-or-map-region-the-r-option

Reply:
> other styles supported by GMT `-R` as mentioned in [#153 (comment)](https://github.com/GenericMappingTools/pygmt/issues/153#issuecomment-742896187) and https://docs.generic-mapping-tools.org/6.4/cookbook/options.html#data-domain-or-map-region-the-r-option

Other styles are more tricky to support. For example, how can we know the boundingbox for `region="US"`?

Reply:
> Other styles are more tricky to support.

Yeah, I was really just thinking about `0/1/2/3`, and maybe `np.array([0, 1, 2, 3])`. Usually users would set a `region` variable near the top of their code, and they might find it strange why their `region` that works for `fig.basemap` doesn't work for `fig.tilemap`.

> For example, how can we know the boundingbox for `region="US"`?

If only there was an API in GMT to return the bounding box for a given region :wink: [Global regions](https://www.pygmt.org/v0.8.0/tutorials/basics/regions.html#global-regions) like `region="g"` and `region="d"` might be easy to support, but ISO code regions (https://www.pygmt.org/v0.8.0/tutorials/basics/regions.html#iso-code) like `region="JP"` and `region=JP+r3" will be a bit trickier.

Reply:
> If only there was an API in GMT to return the bounding box for a given region  [Global regions](https://www.pygmt.org/v0.8.0/tutorials/basics/regions.html#global-regions) like `region="g"` and `region="d"` might be easy to support, but ISO code regions (https://www.pygmt.org/v0.8.0/tutorials/basics/regions.html#iso-code) like `region="JP"` and `region=JP+r3" will be a bit trickier.

Open a upstream feature request in the GMT repository?

Reply:
Hmm, or do you think we can use `GMT_Extract_Region` that's wrapped here:

https://github.com/GenericMappingTools/pygmt/blob/0b5a69b06481335bd4aa7a09f44b779e01e1ce43/pygmt/clib/session.py#L1500-L1571

Downside is that it requires a figure instance to be set up already.

Reply:
I don't think it will work. Actually, a figure instance is not enough. We have to plot something to generate the hidden PostScript file before calling `extract_region`.

```
In [1]: import pygmt

In [2]: fig = pygmt.Figure()

In [3]: fig.region
pygmt-session [ERROR]: No hidden PS file found
---------------------------------------------------------------------------
GMTCLibError                              Traceback (most recent call last)
Cell In[3], line 1
----> 1 fig.region

File ~/OSS/gmt/pygmt/pygmt/figure.py:126, in Figure.region(self)
    124 self._activate_figure()
    125 with Session() as lib:
--> 126     wesn = lib.extract_region()
    127 return wesn

File ~/OSS/gmt/pygmt/pygmt/clib/session.py:1570, in Session.extract_region(self)
   1568 status = c_extract_region(self.session_pointer, None, wesn_pointer)
   1569 if status != 0:
-> 1570     raise GMTCLibError("Failed to extract region from current figure.")
   1571 return wesn
```

Reply:
> I don't think it will work. Actually, a figure instance is not enough. We have to plot something to generate the hidden PostScript file before calling `extract_region`.

Hmm yeah, we would need to maybe do a `fig.basemap` and then `fig.grdimage`, but that's a bit of a hacky solution.

Let's just keep this PR simple and support only a list of `[xmin, xmax, ymin, ymax]` coordinates for now. There can be a follow up PR to support other `region` arguments (most likely after PyGMT v0.9.0) if there's interest in this 'enhancement'.